src/core/lib/iomgr/tcp_client_posix.cc

```diff
@@ -208,6 +241,14 @@ static void on_writable(void* acp, grpc_error_handle error) {
   }
 
 finish:
+  if (!connect_cancelled) {
+    int shard_number = ac->connection_handle % (*g_connection_shards).size();
+    struct ConnectionShard* shard = &(*g_connection_shards)[shard_number];
+    {
+      grpc_core::MutexLock lock(&shard->mu);
```

Comment:
I'm curious about the performance impact of taking another lock on every connection and every write. Have you investigated this?

Reply:
Yeah good point. This lock is not acquired on every write though. The on_writable callback gets executed only once per connection after the connect() succeeds. So the lock will be acquired once before initiating the connection attempt and once after the connection attempt succeeds/fails aynchronously. I dont anticipate contention because we already have a global lock in grpc_fd_create() implementations and haven't observed any major complaints

Reply:
Maybe we can add a flag to control the number of minimum number of shards ?